base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -1888,97 +1891,175 @@ where
+        let wallet_pk = self.node_identity.public_key();
+
+        let mut scanned_outputs = vec![];
 
-        let mut rewound_outputs: Vec<RecoveredOutput> = Vec::new();
         for output in outputs {
-            let position = known_one_sided_payment_scripts
-                .iter()
-                .position(|known_one_sided_script| known_one_sided_script.script == output.script);
-            if let Some(i) = position {
-                let spending_key = PrivateKey::from_bytes(
-                    CommsPublicKey::shared_secret(
-                        &known_one_sided_payment_scripts[i].private_key,
-                        &output.sender_offset_public_key,
+            match output.script.as_slice() {
+                // ----------------------------------------------------------------------------
+                // simple one-sided address
+                [Opcode::PushPubKey(scanned_pk)] => {
+                    match known_keys
+                        .iter()
+                        .find(|x| &PublicKey::from_secret_key(&x.private_key) == scanned_pk.as_ref())
+                    {
+                        // none of the keys match, skipping
+                        None => continue,
+
+                        // match found
+                        Some(matched_key) => {
+                            match PrivateKey::from_bytes(
+                                CommsPublicKey::shared_secret(
+                                    &matched_key.private_key,
+                                    &output.sender_offset_public_key,
+                                )
+                                .as_bytes(),
+                            ) {
+                                Ok(spending_sk) => {
+                                    scanned_outputs.push((output.clone(), matched_key.private_key.clone(), spending_sk))
+                                },
+                                Err(e) => {
+                                    error!(
+                                        target: LOG_TARGET,
+                                        "failed to derive private key from DH shared secret (simple one-sided): {:?}",
+                                        e
+                                    );
+                                    continue;
+                                },
+                            }
+                        },
+                    }
+                },
+
+                // ----------------------------------------------------------------------------
+                // one-sided stealth address
+                // NOTE: Extracting the nonce R and a spending (public aka scan_key) key from the script
+                // NOTE: [RFC 203 on Stealth Addresses](https://rfc.tari.com/RFC-0203_StealthAddresses.html)
+                [Opcode::PushPubKey(nonce), Opcode::Drop, Opcode::PushPubKey(scanned_pk)] => {
+                    // computing shared secret
+                    let c = RistrettoSecretKey::from_bytes(
```

Comment:
```suggestion
                    let stealth_key = PrivateKey::from_bytes(
```
We should use the generic types as far as possible